src/lib/Hydra/Controller/Root.pm

```diff
@@ -442,50 +442,53 @@ sub search :Local Args(0) {
     my $limit = int(trim ($c->request->params->{"limit"} || "10"));
     $c->stash->{limit} = min(50, max(1, $limit));
 
-    $c->stash->{projects} = [ $c->model('DB::Projects')->search(
-        { -and =>
-            [ { -or => [ name => { ilike => "%$query%" }, displayName => { ilike => "%$query%" }, description => { ilike => "%$query%" } ] }
-            , { hidden => 0 }
-            ]
-        },
-        { order_by => ["name"] } ) ];
-
-    $c->stash->{jobsets} = [ $c->model('DB::Jobsets')->search(
-        { -and =>
-            [ { -or => [ "me.name" => { ilike => "%$query%" }, "me.description" => { ilike => "%$query%" } ] }
-            , { "project.hidden" => 0, "me.hidden" => 0 }
-            ]
-        },
-        { order_by => ["project", "name"], join => ["project"] } ) ];
-
-    $c->stash->{jobs} = [ $c->model('DB::Builds')->search(
-        { "job" => { ilike => "%$query%" }
-        , "project.hidden" => 0
-        , "jobset.hidden" => 0
-        , iscurrent => 1
-        },
-        { order_by => ["project", "jobset", "job"], join => ["project", "jobset"]
-        , rows => $c->stash->{limit} + 1
-        } )
-    ];
-
-    # Perform build search in separate queries to prevent seq scan on buildoutputs table.
-    $c->stash->{builds} = [ $c->model('DB::Builds')->search(
-        { "buildoutputs.path" => { ilike => "%$query%" } },
-        { order_by => ["id desc"], join => ["buildoutputs"]
-        , rows => $c->stash->{limit}
-        } ) ];
-
-    $c->stash->{buildsdrv} = [ $c->model('DB::Builds')->search(
-        { "drvpath" => { ilike => "%$query%" } },
-        { order_by => ["id desc"]
-        , rows => $c->stash->{limit}
-        } ) ];
-
-    $c->stash->{resource} = { projects => $c->stash->{projects},
-                              jobsets  => $c->stash->{jobsets},
-                              builds  => $c->stash->{builds},
-                              buildsdrv  => $c->stash->{buildsdrv} };
+    $c->model('DB')->schema->txn_do(sub {
+        $c->model('DB')->schema->storage->dbh->prepare("SET statement_timeout = 20000;")->execute();
```

Comment:
What is the difference compared to `SET LOCAL statement_timeout …`? Does this have the possibility to escape the transaction?

Reply:
According to https://www.postgresql.org/docs/10/sql-set.html this would set the timeout for all other queries, even those outside of the transaction. We probably want `SET LOCAL` instead:

```
The effects of SET LOCAL last only till the end of the current transaction, whether committed or not. A special case is SET followed by SET LOCAL within a single transaction: the SET LOCAL value will be seen until the end of the transaction, but afterwards (if the transaction is committed) the SET value will take effect.
```

Reply:
Good catch.

> If SET (or equivalently SET SESSION) is issued within a transaction that is later aborted, the effects of the SET command disappear when the transaction is rolled back. Once the surrounding transaction is committed, the effects will persist until the end of the session, unless overridden by another SET.
>
>The effects of SET LOCAL last only till the end of the current transaction, whether committed or not. A special case is SET followed by SET LOCAL within a single transaction: the SET LOCAL value will be seen until the end of the transaction, but afterwards (if the transaction is committed) the SET value will take effect.

Changing to SET LOCAL.

Reply:
Fixed, tested locally that it still works. Thanks!